store/src/java/com/zimbra/cs/service/admin/CreateVolume.java

```diff
@@ -51,35 +51,45 @@ private CreateVolumeResponse handle(CreateVolumeRequest req, Map<String, Object>
         checkRight(zsc, ctx, Provisioning.getInstance().getLocalServer(), Admin.R_manageVolume);
 
         VolumeInfo volInfoRequest = req.getVolume();
-        Volume volRequest = VolumeManager.getInstance().create(toVolume(volInfoRequest));
-        VolumeInfo volInfoResponse = volRequest.toJAXB();
+        Volume.StoreType enumStoreType = (1 == volInfoRequest.getStoreType()) ? Volume.StoreType.INTERNAL : Volume.StoreType.EXTERNAL;
+        ExternalVolumeInfoHandler extVolInfoHandler = new ExternalVolumeInfoHandler(Provisioning.getInstance());
 
-        // if newly created volume is external update json
-        if (volRequest.getStoreType().equals(Volume.StoreType.EXTERNAL)) {
-            Provisioning prov = Provisioning.getInstance();
+        if (enumStoreType.equals(Volume.StoreType.EXTERNAL)) {
             try {
-                // As id is created once volume is created, update id
-                volInfoRequest.setId(volRequest.getId());
+                String globalS3BucketId = volInfoRequest.getVolumeExternalInfo().getGlobalBucketConfigurationId();
+                if (false == extVolInfoHandler.validateGlobalBucketID(globalS3BucketId)) {
+                    ZimbraLog.store.error("Invalid global bucket ID provided %s", globalS3BucketId);
+                    throw ServiceException.FAILURE("Invalid global bucket ID provided", null);
+                }
+            } catch (ServiceException e) {
+                ZimbraLog.store.error("Error while processing CreateVolumeRequest", e);
+                throw e;
+            } catch (JSONException e) {
+                throw ServiceException.FAILURE("Error while processing CreateVolumeRequest", e);
+            }
+        }
 
-                // Update JSON state server properties
-                ExternalVolumeInfoHandler extVolInfoHandler = new ExternalVolumeInfoHandler(prov);
-                extVolInfoHandler.addServerProperties(volInfoRequest);
+        // create volume response
+        Volume volRequest = VolumeManager.getInstance().create(toVolume(volInfoRequest, enumStoreType));
+        VolumeInfo volInfoResponse = volRequest.toJAXB();
 
-                // Add External volume info to response
+        if (enumStoreType.equals(Volume.StoreType.EXTERNAL)) {
+            try {
+                volInfoRequest.setId(volRequest.getId());
                 volInfoResponse.setVolumeExternalInfo(volInfoRequest.getVolumeExternalInfo());
+                extVolInfoHandler.addServerProperties(volInfoRequest);
             } catch (ServiceException e) {
                 ZimbraLog.store.error("Error while processing CreateVolumeRequest", e);
                 throw e;
             } catch  (JSONException e) {
```

Comment:
I see one problem here. Consider the case, Volume info is inserted in the table but it fails to insert external info in the JSON object, you are throwing an error but keeping volume stale entry database.

Reply:
We will just throw error for now